### PR TITLE
nip96: upload file

### PR DESCRIPTION
**Pull request by bob2402**

### CODE REVIEW

File: event.ts [hunk 1]
```diff
@@ -74,3 +74,34 @@ export async function prepareDeletionEvent(
         },
     );
 }
+
+export async function prepareHttpAuthEvent(
```
Comment (BlowaterNostr) [inline event.ts:78]: let's move to `nip96.ts`

File: nip96.test.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import { assertEquals } from "https://deno.land/std@0.202.0/assert/assert_equals.ts";
+import { uploadFile } from "./nip96.ts";
+import { InMemoryAccountContext } from "./nostr.ts";
+
+Deno.test("Upload File", async () => {
+    const ctx = InMemoryAccountContext.Generate();
+    const url = "https://nostr.build/api/v2/nip96/upload";
+    const file = new Blob([await Deno.readFile("./test.jpeg")], { type: "image/jpeg" });
+    const result = await uploadFile(ctx, { url, file });
+    console.log(result);
```
Comment (BlowaterNostr) [inline nip96.test.ts:10]: assert the result instead of logging

File: nip96.test.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import { assertEquals } from "https://deno.land/std@0.202.0/assert/assert_equals.ts";
+import { uploadFile } from "./nip96.ts";
+import { InMemoryAccountContext } from "./nostr.ts";
+
+Deno.test("Upload File", async () => {
+    const ctx = InMemoryAccountContext.Generate();
+    const url = "https://nostr.build/api/v2/nip96/upload";
+    const file = new Blob([await Deno.readFile("./test.jpeg")], { type: "image/jpeg" });
```
Comment (BlowaterNostr) [inline nip96.test.ts:8]: Let's use an image in the internet instead of committing a file to the git.

File: nip96.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    console.log("httpAuthEvent", httpAuthEvent);
+
+    const result = await fetch(args.url, {
+        method: "POST",
+        headers: {
+            "Authorization": `Nostr ${btoa(JSON.stringify(httpAuthEvent))}`,
+        },
+        body: formData,
+    });
+    return result.json() as Promise<UploadFileResponse>;
```
Comment (BlowaterNostr) [inline nip96.ts:41]: Does this endpoint always return 200? If not, need to do traditional status code handling

File: nostr.ts [hunk 5]
```diff
@@ -16,6 +16,7 @@ export enum NostrKind {
     Custom_App_Data = 30078, // https://github.com/nostr-protocol/nips/blob/master/78.md
     Long_Form = 30023, // https://github.com/nostr-protocol/nips/blob/master/23.md
     Group_Message = 1000,
```
Comment (BlowaterNostr) [inline nostr.ts:18]: let's remove this as well, not used anymore

File: nip96.ts (hunk 4 above)
Comment (BlowaterNostr) [reply nip96.ts:41]: JSON parse could fail and throw as well.

**Review by BlowaterNostr — CHANGES_REQUESTED** [review]

File: makefile [hunk 6]
```diff
@@ -1,6 +1,6 @@
 test: fmt
 	rm -rf cov_profile*
-	deno test --allow-net --trace-ops --coverage=cov_profile *.test.ts
+	deno test --allow-net --trace-leaks --coverage=cov_profile *.test.ts
```
Comment (BlowaterNostr) [inline makefile:3]: keep both

File: relayed.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import { run } from "https://raw.githubusercontent.com/BlowaterNostr/relayed/main/main.tsx";
+
+const relay = await run({
+    port: 8080,
+    default_policy: {
+        allowed_kinds: "all", // or none,
+    },
+    system_key: "nsec1geplx78d053a7554j8xtrfgffknukmyc6w206r2dszdsl2t2zhnqw0g3fg",
```
Comment (BlowaterNostr) [inline relayed.ts:8]: Why do we need a hard-coded nsec?

File: relayed.ts
Comment (BlowaterNostr) [inline relayed.ts:1]: Rename to `relayed.test.ts` or `test.relayed.ts` to indicate that it's not a production module

**Review by BlowaterNostr — CHANGES_REQUESTED** [review]

File: relay-single-test.ts [hunk 8]
```diff
@@ -177,9 +173,21 @@ export const limit = (url: string) => async () => {
             }),
         );
         if (err3 instanceof Error) fail(err3.message);
+        const err4 = await relay.sendEvent(
+            await prepareNormalNostrEvent(ctx, {
+                kind: NostrKind.TEXT_NOTE,
+                content: "4",
+            }),
+        );
+        if (err4 instanceof Error) fail(err4.message);
```
[the remark this answers is not in the document]
Comment (zhengbangbo) [reply relay-single-test.ts:182]: Send 4 but receive only 3 to test its limit.

[the remark this answers is not in the document]
Comment (bob2402) [reply relay-single-test.ts:182]: Send 4 but receive only 3 to test its limit.

**Review by BlowaterNostr — APPROVED** [review]

**Pull request merged**